README.md
# atom-live-server package

Launch a simple development http server with live reload capability.

![Demo](https://raw.githubusercontent.com/jas-chen/atom-live-server/master/doc/demo.gif)

This package is based on awesome [Live Server](https://github.com/tapio/live-server) project.

## Usage

`ctrl-alt-l` launch live server on port 3000.

`ctrl-alt-q` stop live server.

`ctrl-alt-3` launch live server on port 3000.

`ctrl-alt-4` launch live server on port 4000.

`ctrl-alt-5` launch live server on port 5000.

`ctrl-alt-8` launch live server on port 8000.

`ctrl-alt-9` launch live server on port 9000.


## Options

If a file `.atom-live-server.json` exists in project root it will be loaded and used as options.

The keys of `.atom-live-server.json` should match **Command line parameters** in [live-server](https://github.com/tapio/live-server).

#### Example: Launch specific browser
```json
{
  "browser": "safari"
}
```

#### Example: Start server without browser opened
```json
{
  "no-browser": true
}
```

#### Example: Serve this file for every 404 (useful for single-page applications)
```json
{
  "entry-file": "index.html"
}
```


CHANGELOG.md
## 2.0.0

#### New features
* Live server can be stopped.
* Support `.atom-live-server.json`.

#### Breaking changes
* Keymaps changed


spec/atom-live-server-spec.js
'use babel';

import AtomLiveServer from '../lib/atom-live-server';

// Use the command `window:run-package-specs` (cmd-alt-ctrl-p) to run specs.
//
// To run a specific `it` or `describe` block add an `f` to the front (e.g. `fit`
// or `fdescribe`). Remove the `f` to unfocus the block.

describe('AtomLiveServer', () => {
  let workspaceElement, activationPromise;

  beforeEach(() => {
    workspaceElement = atom.views.getView(atom.workspace);
    activationPromise = atom.packages.activatePackage('atom-live-server');
  });

  describe('when the atom-live-server:toggle event is triggered', () => {
    it('hides and shows the modal panel', () => {
      // Before the activation event the view is not on the DOM, and no panel
      // has been created
      expect(workspaceElement.querySelector('.atom-live-server')).not.toExist();

      // This is an activation event, triggering it will cause the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'atom-live-server:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        expect(workspaceElement.querySelector('.atom-live-server')).toExist();

        let atomLiveServerElement = workspaceElement.querySelector('.atom-live-server');
        expect(atomLiveServerElement).toExist();

        let atomLiveServerPanel = atom.workspace.panelForItem(atomLiveServerElement);
        expect(atomLiveServerPanel.isVisible()).toBe(true);
        atom.commands.dispatch(workspaceElement, 'atom-live-server:toggle');
        expect(atomLiveServerPanel.isVisible()).toBe(false);
      });
    });

    it('hides and shows the view', () => {
      // This test shows you an integration test testing at the view level.

      // Attaching the workspaceElement to the DOM is required to allow the
      // `toBeVisible()` matchers to work. Anything testing visibility or focus
      // requires that the workspaceElement is on the DOM. Tests that attach the
      // workspaceElement to the DOM are generally slower than those off DOM.
      jasmine.attachToDOM(workspaceElement);

      expect(workspaceElement.querySelector('.atom-live-server')).not.toExist();

      // This is an activation event, triggering it causes the package to be
      // activated.
      atom.commands.dispatch(workspaceElement, 'atom-live-server:toggle');

      waitsForPromise(() => {
        return activationPromise;
      });

      runs(() => {
        // Now we can test for view visibility
        let atomLiveServerElement = workspaceElement.querySelector('.atom-live-server');
        expect(atomLiveServerElement).toBeVisible();
        atom.commands.dispatch(workspaceElement, 'atom-live-server:toggle');
        expect(atomLiveServerElement).not.toBeVisible();
      });
    });
  });
});


spec/atom-live-server-view-spec.js
'use babel';

import AtomLiveServerView from '../lib/atom-live-server-view';

describe('AtomLiveServerView', () => {
  it('has one valid test', () => {
    expect('life').toBe('easy');
  });
});


live-server/index.js
#!/usr/bin/env node
var fs = require('fs'),
	connect = require('connect'),
	colors = require('colors'),
	WebSocket = require('faye-websocket'),
	path = require('path'),
	url = require('url'),
	http = require('http'),
	send = require('send'),
	open = require('open'),
	es = require("event-stream"),
	watchr = require('watchr');

var INJECTED_CODE = fs.readFileSync(__dirname + "/injected.html", "utf8");

var LiveServer = {};

function escape(html){
	return String(html)
		.replace(/&(?!\w+;)/g, '&amp;')
		.replace(/</g, '&lt;')
		.replace(/>/g, '&gt;')
		.replace(/"/g, '&quot;');
}

// Based on connect.static(), but streamlined and with added code injecter
function staticServer(root) {
	return function(req, res, next) {
		if ('GET' != req.method && 'HEAD' != req.method) return next();
		var reqpath = url.parse(req.url).pathname;
		var hasNoOrigin = !req.headers.origin;
		var doInject = false;

		function directory() {
			var pathname = url.parse(req.originalUrl).pathname;
			res.statusCode = 301;
			res.setHeader('Location', pathname + '/');
			res.end('Redirecting to ' + escape(pathname) + '/');
		}

		function file(filepath, stat) {
			var x = path.extname(filepath);
			if (hasNoOrigin && (x === "" || x == ".html" || x == ".htm" || x == ".xhtml" || x == ".php")) {
				// TODO: Sync file read here is not nice, but we need to determine if the html should be injected or not
				var contents = fs.readFileSync(filepath, "utf8");
				doInject = contents.indexOf("</body>") > -1;
			}
		}

		function error(err) {
			if (404 == err.status) return next();
			next(err);
		}

		function inject(stream) {
			if (doInject) {
				// We need to modify the length given to browser
				var len = INJECTED_CODE.length + res.getHeader('Content-Length');
				res.setHeader('Content-Length', len);
				var originalPipe = stream.pipe;
				stream.pipe = function(res) {
					originalPipe.call(stream, es.replace(new RegExp("</body>","i"), INJECTED_CODE + "</body>")).pipe(res);
				};
			}
		}

		send(req, reqpath, { root: root })
			.on('error', error)
			.on('directory', directory)
			.on('file', file)
			.on('stream', inject)
			.pipe(res);
	};
}

/**
 * Rewrite request URL and pass it back to the static handler.
 * @param staticHandler {function} Next handler
 * @param file {string} Path to the entry point file
 */
function entryPoint(staticHandler, file) {
	if (!file) return function(req, res, next) { next(); };

	return function(req, res, next) {
		req.url = "/" + file;
		staticHandler(req, res, next);
	};
}

/**
 * Start a live server with parameters given as an object
 * @param host {string} Address to bind to (default: 0.0.0.0)
 * @param port {number} Port number (default: 8080)
 * @param root {string} Path to root directory (default: cwd)
 * @param open {string} Subpath to open in browser, use false to suppress launch (default: server root)
 * @param logLevel {number} 0 = errors only, 1 = some, 2 = lots
 * @param file {string} Path to the entry point file
 * @param wait {number} Server will wait for all changes, before reloading
 */
LiveServer.start = function(options) {
	options = options || {};
	var host = options.host || '0.0.0.0';
	var port = options.port || 8080;
	var root = options.root || process.cwd();
	var logLevel = options.logLevel === undefined ? 2 : options.logLevel;
	var openPath = (options.open === undefined || options.open === true) ?
		"" : ((options.open === null || options.open === false) ? null : options.open);
	if (options.noBrowser) openPath = null; // Backwards compatibility with 0.7.0
	var file = options.file;
	var staticServerHandler = staticServer(root);
	var wait = options.wait || 0;
	var browser = options.browser;

	// Setup a web server
	var app = connect()
		.use(staticServerHandler) // Custom static server
		.use(entryPoint(staticServerHandler, file))
		.use(connect.directory(root, { icons: true }));
	if (logLevel >= 2)
		app.use(connect.logger('dev'));
	var server = http.createServer(app);

	// Handle server startup errors
	server.addListener('error', function(e) {
		if (e.code == 'EADDRINUSE') {
			var serveURL = 'http://' + host + ':' +  port;
			console.log('%s is already in use. Trying another port.'.red, serveURL);
			setTimeout(function() {
				server.listen(0, host);
			}, 1000);
		}
	});

	// Handle successful server
	server.addListener('listening', function(e) {
		var address = server.address();
		var serveHost = address.address == "0.0.0.0" ? "127.0.0.1" : address.address;
		var serveURL = 'http://' + serveHost + ':' +  address.port;

		// Output
		if (logLevel >= 1) {
			console.log(("Serving \"%s\" at %s").green, root, serveURL);
		}

		// Launch browser
		if (openPath !== null) {
			if (browser === 'system' || browser === '') {
				open(serveURL + openPath);
			} else {
				open(serveURL + openPath, browser);
			}
		}
	});

	// Setup server to listen at port
	server.listen(port, host);

	// WebSocket
	var clients = [];
	server.addListener('upgrade', function(request, socket, head) {
		var ws = new WebSocket(request, socket, head);
		ws.onopen = function() { ws.send('connected'); };

		if (wait > 0) {
			(function(ws) {
				var wssend = ws.send;
				var waitTimeout;

				ws.send = function() {
					var args = arguments;
					if (waitTimeout) clearTimeout(waitTimeout);
					waitTimeout = setTimeout(function(){
						wssend.apply(ws, args);
					}, wait);
				};
			})(ws);
		}

		ws.onclose = function() {
			clients = clients.filter(function (x) {
				return x !== ws;
			});
		};

		clients.push(ws);
	});

	// Setup file watcher
	watchr.watch({
		path: root,
		ignorePaths: options.ignore || false,
		ignoreCommonPatterns: true,
		ignoreHiddenFiles: true,
		preferredMethods: [ 'watchFile', 'watch' ],
		interval: 1407,
		listeners: {
			error: function(err) {
				console.log("ERROR:".red , err);
			},
			change: function(eventName, filePath, fileCurrentStat, filePreviousStat) {
				clients.forEach(function(ws) {
					if (!ws) return;
					if (path.extname(filePath) == ".css") {
						ws.send('refreshcss');
						if (logLevel >= 1)
							console.log("CSS change detected".magenta);
					} else {
						ws.send('reload');
						if (logLevel >= 1)
							console.log("File change detected".cyan);
					}
				});
			}
		}
	});
};

module.exports = LiveServer;


lib/atom-live-server.js
'use babel';

import path from 'path';
import fs from 'fs';
import url from 'url';
import { BufferedProcess, CompositeDisposable } from 'atom';
import { remote } from 'electron';
import JSON5 from 'json5';

const packagePath = atom.packages.resolvePackagePath('atom-live-server');
const liveServer = path.join(packagePath, '/node_modules/live-server/live-server.js');
const node = path.resolve(process.env.NODE_PATH, '../../app/apm/bin/node');

let serverProcess;
let disposeMenu;
let noBrowser;

function addStartMenu() {
  disposeMenu = atom.menu.add(
    [{
      label: 'Packages',
      submenu : [{
        label: 'atom-live-server',
        submenu : [{
          label: 'Start server',
          command: `atom-live-server:startServer`
        }]
      }]
    }]
  );
}

export default {
  subscriptions: null,

  activate(state) {
    this.subscriptions = new CompositeDisposable();

    this.subscriptions.add(atom.commands.add('atom-workspace', {
      'atom-live-server:start-3000': () => this.startServer(3000),
      'atom-live-server:start-4000': () => this.startServer(4000),
      'atom-live-server:start-5000': () => this.startServer(5000),
      'atom-live-server:start-8000': () => this.startServer(8000),
      'atom-live-server:start-9000': () => this.startServer(9000),
      'atom-live-server:startServer': () => this.startServer(),
      'atom-live-server:stopServer': () => this.stopServer()
    }));

    addStartMenu();
  },

  deactivate() {
    this.stopServer();
    this.subscriptions.dispose();
  },

  startServer(port = 3000) {
    if (serverProcess) {
      return;
    }

    const targetPath = atom.project.getPaths()[0];

    if (!targetPath) {
      atom.notifications.addWarning('[Live Server] You haven\'t opened a Project, you must open one.')
      return;
    }

    noBrowser = false;
    const args = [];
    const stdout = output => {
      if (output.indexOf('Serving ') === 0) {
        const serverUrl = output.split(' at ')[1];
        const port = url.parse(serverUrl).port;
        const disposeStartMenu = disposeMenu;
        disposeMenu = atom.menu.add(
          [{
            label: 'Packages',
            submenu : [{
              label: 'atom-live-server',
              submenu : [{
                label: output.replace('Serving ', 'Stop '),
                command: `atom-live-server:stopServer`
              }]
            }]
          }]
        );

        disposeStartMenu.dispose();

        if (noBrowser) {
          atom.notifications.addSuccess(`[Live Server] Live server started at ${serverUrl}.`);
        }
      }

      console.log(output);
    };

    const exit = code => console.log(`live-server exited with code ${code}`);

    fs.open(path.join(targetPath, '.atom-live-server.json'), 'r', (err, fd) => {
      if (!err) {
        const userConfig = JSON5.parse(fs.readFileSync(fd, 'utf8'));

        Object.keys(userConfig).forEach(key => {
          if (key === 'no-browser') {
            if (userConfig[key] === true) {
              args.push(`--${key}`);
              noBrowser = true;
            }
          } else {
            args.push(`--${key}=${userConfig[key]}`);
          }
        });
      }

      if (!args.length) {
        args.push(`--port=${port}`);
      }

      args.unshift(liveServer);

      serverProcess = new BufferedProcess({
        command: node,
        args,
        stdout,
        exit,
        options: {
          cwd: targetPath,
        }
      });

      console.info(`live-server ${args.join(' ')}`);
    });
  },

  stopServer() {
    if (serverProcess) {
      console.info('Stopping live-server.');
      serverProcess.kill();
      serverProcess = null;
      const disposeStopMenu = disposeMenu;
      addStartMenu();
      disposeStopMenu.dispose();
      atom.notifications.addSuccess('[Live Server] Live server is stopped.')
    }
  }
};
